packages/beacon-node/src/api/impl/validator/index.ts

```diff
@@ -926,21 +931,59 @@ export function getValidatorApi(
       }
 
       if (!state) {
-        state = await chain.getHeadStateAtCurrentEpoch(RegenCaller.getDuties);
+        if (epoch >= currentEpoch - 1) {
+          // Cached beacon state stores proposers for previous, current and next epoch. The
+          // requested epoch is within that range, we can use the head state at current epoch
+          state = await chain.getHeadStateAtCurrentEpoch(RegenCaller.getDuties);
+        } else {
+          const res = await getStateResponseWithRegen(chain, startSlot);
+
+          const stateViewDU =
+            res.state instanceof Uint8Array
+              ? loadState(config, chain.getHeadState(), res.state).state
```

Comment:
we should make sure the loaded state has the same epoch to the `epoch` param

Reply:
This seems really unexpected as we get state for start slot of the epoch
https://github.com/ChainSafe/lodestar/blob/ea67800e1b950e1b46aff0b96a001d6ff9a5b7b3/packages/beacon-node/src/api/impl/validator/index.ts#L939

can you elaborate why this can happen, or do we just wanna add an assert to make sure we catch downstream bugs?

Reply:
To add to that, there was a bug in historical state regen where it would return a state for different slot but that's asserted now and will throw https://github.com/ChainSafe/lodestar/pull/7135

We also have a sanity check later in the flow
https://github.com/ChainSafe/lodestar/blob/ea67800e1b950e1b46aff0b96a001d6ff9a5b7b3/packages/beacon-node/src/api/impl/validator/index.ts#L985-L986

Reply:
> can you elaborate why this can happen, or do we just wanna add an assert to make sure we catch downstream bugs?

I just want to assert, to make sure the below assumption  correct `and if we load state from `Uint8Array` it will always be the state of requested epoch`

if a bug happen because of that, we can easily catch it

Reply:
yeah, I think having many assertions in the code is a good thing, let me add that